{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e8e2944",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e4c3e41",
   "metadata": {},
   "outputs": [],
   "source": [
    "projdir = '/u/home/t/terencew/project-cluo/igvf/pilot/multiome/'\n",
    "\n",
    "donors = list(np.loadtxt(f'{projdir}/txt/donors.txt', dtype=str))\n",
    "samples = list(np.loadtxt(f'{projdir}/txt/samples.txt', dtype=str))\n",
    "s = samples[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ccfe3921",
   "metadata": {},
   "outputs": [],
   "source": [
    "doub_experiments = np.loadtxt(f'{projdir}/ambient/ambisim/cov_doub_test/txt/experiments.txt', dtype=str)\n",
    "mux_experiments = np.loadtxt(f'{projdir}/ambient/ambisim/cov_mux_test/txt/experiments.txt', dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d67a82bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_truth_df(indir, crarc_dir):\n",
    "    truth = pd.read_csv(f'{indir}/drop_data_rand.txt', sep='\\t', header=0, index_col=0, low_memory=False)\n",
    "    barcodes = pd.read_csv(f'{crarc_dir}/outs/filtered_feature_bc_matrix/barcodes.tsv.gz',\n",
    "                       sep='\\t', header=None, index_col=None)\n",
    "\n",
    "    truth.index = [f'{x}-1' for x in truth.index]\n",
    "    truth = truth[truth.index.isin(barcodes[0])]\n",
    "    truth.sort_index(inplace=True)\n",
    "   \n",
    "    ###\n",
    "    doub_mask = [',' in x for x in truth['sam']]\n",
    "    truth.loc[doub_mask, 'sam'] = 'doublet'\n",
    "        \n",
    "    truth['ambient_RNA'] = truth['rna_nr_a'] / (truth['rna_nr_c']  + truth['rna_nr_a'])\n",
    "    truth['ambient_DNA'] = truth['atac_nr_a'] / (truth['atac_nr_c']  + truth['atac_nr_a'])\n",
    "    truth['ambient_DNA_peaks'] = truth['atac_nr_ap'] / (truth['atac_nr_cp']  + truth['atac_nr_ap'])\n",
    "    truth['method'] = 'truth'\n",
    "    \n",
    "    truth = truth[['sam', 'ambient_RNA', 'ambient_DNA', 'ambient_DNA_peaks', 'method']]\n",
    "    truth.rename(columns={'sam' : 'donor_id'}, inplace=True)\n",
    "\n",
    "    return truth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9992934e",
   "metadata": {},
   "outputs": [],
   "source": [
    "indir = f'{projdir}/ambient/ambisim/cov_doub_test/'\n",
    "doub_summary = pd.DataFrame()\n",
    "for exp in doub_experiments:\n",
    "    tmp_summary = pd.read_csv(f'{indir}/{exp}/cr_arc/{s}/outs/summary.csv', sep=',')\n",
    "    tmp_summary.index = [exp]\n",
    "    doub_summary = pd.concat([doub_summary, tmp_summary])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a8c8fc9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "indir = f'{projdir}/ambient/ambisim/cov_mux_test/'\n",
    "mux_summary = pd.DataFrame()\n",
    "for exp in mux_experiments:\n",
    "    tmp_summary = pd.read_csv(f'{indir}/{exp}/cr_arc/{s}/outs/summary.csv', sep=',')\n",
    "    tmp_summary.index = [exp]\n",
    "    mux_summary = pd.concat([mux_summary, tmp_summary])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa5f47f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "doub_summary['type'] = 'Vary_doublet'\n",
    "mux_summary['type'] = 'Vary_num_muxed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f130f480",
   "metadata": {},
   "outputs": [],
   "source": [
    "doub_summary['amb_rate'] = [x.split('_')[0] for x in doub_summary.index]\n",
    "doub_summary['doublet_rate'] = [x.split('_')[1] for x in doub_summary.index]\n",
    "doub_summary['num_donors'] = [4 for x in doub_summary.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a46ebe75",
   "metadata": {},
   "outputs": [],
   "source": [
    "mux_summary['amb_rate'] = [x.split('_')[1] for x in mux_summary.index]\n",
    "mux_summary['doublet_rate'] = [10 for x in mux_summary.index]\n",
    "mux_summary['num_donors'] = [x.split('_')[0] for x in mux_summary.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ed56be53",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_summary = pd.concat([doub_summary, mux_summary])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "00e7e42b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Pipeline version</th>\n",
       "      <th>Estimated number of cells</th>\n",
       "      <th>Feature linkages detected</th>\n",
       "      <th>Linked genes</th>\n",
       "      <th>Linked peaks</th>\n",
       "      <th>ATAC Confidently mapped read pairs</th>\n",
       "      <th>ATAC Fraction of genome in peaks</th>\n",
       "      <th>ATAC Fraction of high-quality fragments in cells</th>\n",
       "      <th>...</th>\n",
       "      <th>GEX Reads mapped to genome</th>\n",
       "      <th>GEX Reads with TSO</th>\n",
       "      <th>GEX Sequenced read pairs</th>\n",
       "      <th>GEX Total genes detected</th>\n",
       "      <th>GEX Valid UMIs</th>\n",
       "      <th>GEX Valid barcodes</th>\n",
       "      <th>type</th>\n",
       "      <th>amb_rate</th>\n",
       "      <th>doublet_rate</th>\n",
       "      <th>num_donors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0_0</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8631</td>\n",
       "      <td>15570</td>\n",
       "      <td>5270</td>\n",
       "      <td>11303</td>\n",
       "      <td>0.8748</td>\n",
       "      <td>0.0424</td>\n",
       "      <td>0.9335</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60314841</td>\n",
       "      <td>29202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_10</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8584</td>\n",
       "      <td>13139</td>\n",
       "      <td>4687</td>\n",
       "      <td>10368</td>\n",
       "      <td>0.8744</td>\n",
       "      <td>0.0424</td>\n",
       "      <td>0.9264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9980</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60131721</td>\n",
       "      <td>29242</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_20</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8571</td>\n",
       "      <td>11718</td>\n",
       "      <td>4469</td>\n",
       "      <td>9926</td>\n",
       "      <td>0.8749</td>\n",
       "      <td>0.0425</td>\n",
       "      <td>0.9246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59914589</td>\n",
       "      <td>29164</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_30</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8572</td>\n",
       "      <td>10058</td>\n",
       "      <td>3964</td>\n",
       "      <td>9226</td>\n",
       "      <td>0.8748</td>\n",
       "      <td>0.0427</td>\n",
       "      <td>0.9224</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60091431</td>\n",
       "      <td>29190</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_0</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8594</td>\n",
       "      <td>11004</td>\n",
       "      <td>4241</td>\n",
       "      <td>8788</td>\n",
       "      <td>0.8710</td>\n",
       "      <td>0.0401</td>\n",
       "      <td>0.9264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60013084</td>\n",
       "      <td>29121</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID  Genome      Pipeline version  \\\n",
       "0_0   20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_10  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_20  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_30  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "10_0  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "\n",
       "      Estimated number of cells  Feature linkages detected  Linked genes  \\\n",
       "0_0                        8631                      15570          5270   \n",
       "0_10                       8584                      13139          4687   \n",
       "0_20                       8571                      11718          4469   \n",
       "0_30                       8572                      10058          3964   \n",
       "10_0                       8594                      11004          4241   \n",
       "\n",
       "      Linked peaks  ATAC Confidently mapped read pairs  \\\n",
       "0_0          11303                              0.8748   \n",
       "0_10         10368                              0.8744   \n",
       "0_20          9926                              0.8749   \n",
       "0_30          9226                              0.8748   \n",
       "10_0          8788                              0.8710   \n",
       "\n",
       "      ATAC Fraction of genome in peaks  \\\n",
       "0_0                             0.0424   \n",
       "0_10                            0.0424   \n",
       "0_20                            0.0425   \n",
       "0_30                            0.0427   \n",
       "10_0                            0.0401   \n",
       "\n",
       "      ATAC Fraction of high-quality fragments in cells  ...  \\\n",
       "0_0                                             0.9335  ...   \n",
       "0_10                                            0.9264  ...   \n",
       "0_20                                            0.9246  ...   \n",
       "0_30                                            0.9224  ...   \n",
       "10_0                                            0.9264  ...   \n",
       "\n",
       "      GEX Reads mapped to genome  GEX Reads with TSO  \\\n",
       "0_0                       0.9981                 0.0   \n",
       "0_10                      0.9980                 0.0   \n",
       "0_20                      0.9981                 0.0   \n",
       "0_30                      0.9981                 0.0   \n",
       "10_0                      0.9981                 0.0   \n",
       "\n",
       "      GEX Sequenced read pairs  GEX Total genes detected  GEX Valid UMIs  \\\n",
       "0_0                   60314841                     29202             1.0   \n",
       "0_10                  60131721                     29242             1.0   \n",
       "0_20                  59914589                     29164             1.0   \n",
       "0_30                  60091431                     29190             1.0   \n",
       "10_0                  60013084                     29121             1.0   \n",
       "\n",
       "      GEX Valid barcodes          type  amb_rate  doublet_rate  num_donors  \n",
       "0_0                  1.0  Vary_doublet         0             0           4  \n",
       "0_10                 1.0  Vary_doublet         0            10           4  \n",
       "0_20                 1.0  Vary_doublet         0            20           4  \n",
       "0_30                 1.0  Vary_doublet         0            30           4  \n",
       "10_0                 1.0  Vary_doublet        10             0           4  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d1429627",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_summary.to_csv(f'{projdir}/csv/final_figures/tables/table_s3_sims_crarc_qc.csv',\n",
    "                   sep=',', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "618c6398",
   "metadata": {},
   "source": [
    "### lowcov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "187c24ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "indir = f'{projdir}/ambient/ambisim/prop_doub/'\n",
    "doub_summary = pd.DataFrame()\n",
    "for exp in doub_experiments:\n",
    "    tmp_summary = pd.read_csv(f'{indir}/{exp}/cr_arc/{s}/outs/summary.csv', sep=',')\n",
    "    tmp_summary.index = [exp]\n",
    "    doub_summary = pd.concat([doub_summary, tmp_summary])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a174beb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "indir = f'{projdir}/ambient/ambisim/mux_test/'\n",
    "mux_summary = pd.DataFrame()\n",
    "for exp in mux_experiments:\n",
    "    tmp_summary = pd.read_csv(f'{indir}/{exp}/cr_arc/{s}/outs/summary.csv', sep=',')\n",
    "    tmp_summary.index = [exp]\n",
    "    mux_summary = pd.concat([mux_summary, tmp_summary])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8fe97f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "doub_summary['amb_rate'] = [x.split('_')[0] for x in doub_summary.index]\n",
    "doub_summary['doublet_rate'] = [x.split('_')[1] for x in doub_summary.index]\n",
    "doub_summary['num_donors'] = [4 for x in doub_summary.index]\n",
    "mux_summary['amb_rate'] = [x.split('_')[1] for x in mux_summary.index]\n",
    "mux_summary['doublet_rate'] = [10 for x in mux_summary.index]\n",
    "mux_summary['num_donors'] = [x.split('_')[0] for x in mux_summary.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2fafda69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Pipeline version</th>\n",
       "      <th>Estimated number of cells</th>\n",
       "      <th>Feature linkages detected</th>\n",
       "      <th>Linked genes</th>\n",
       "      <th>Linked peaks</th>\n",
       "      <th>ATAC Confidently mapped read pairs</th>\n",
       "      <th>ATAC Fraction of genome in peaks</th>\n",
       "      <th>ATAC Fraction of high-quality fragments in cells</th>\n",
       "      <th>...</th>\n",
       "      <th>GEX Reads mapped to genome</th>\n",
       "      <th>GEX Reads with TSO</th>\n",
       "      <th>GEX Sequenced read pairs</th>\n",
       "      <th>GEX Total genes detected</th>\n",
       "      <th>GEX Valid UMIs</th>\n",
       "      <th>GEX Valid barcodes</th>\n",
       "      <th>type</th>\n",
       "      <th>amb_rate</th>\n",
       "      <th>doublet_rate</th>\n",
       "      <th>num_donors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0_0</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8631</td>\n",
       "      <td>15570</td>\n",
       "      <td>5270</td>\n",
       "      <td>11303</td>\n",
       "      <td>0.8748</td>\n",
       "      <td>0.0424</td>\n",
       "      <td>0.9335</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60314841</td>\n",
       "      <td>29202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_10</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8584</td>\n",
       "      <td>13139</td>\n",
       "      <td>4687</td>\n",
       "      <td>10368</td>\n",
       "      <td>0.8744</td>\n",
       "      <td>0.0424</td>\n",
       "      <td>0.9264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9980</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60131721</td>\n",
       "      <td>29242</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_20</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8571</td>\n",
       "      <td>11718</td>\n",
       "      <td>4469</td>\n",
       "      <td>9926</td>\n",
       "      <td>0.8749</td>\n",
       "      <td>0.0425</td>\n",
       "      <td>0.9246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59914589</td>\n",
       "      <td>29164</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_30</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8572</td>\n",
       "      <td>10058</td>\n",
       "      <td>3964</td>\n",
       "      <td>9226</td>\n",
       "      <td>0.8748</td>\n",
       "      <td>0.0427</td>\n",
       "      <td>0.9224</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60091431</td>\n",
       "      <td>29190</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_0</th>\n",
       "      <td>20220928-IGVF-D0</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>cellranger-arc-2.0.1</td>\n",
       "      <td>8594</td>\n",
       "      <td>11004</td>\n",
       "      <td>4241</td>\n",
       "      <td>8788</td>\n",
       "      <td>0.8710</td>\n",
       "      <td>0.0401</td>\n",
       "      <td>0.9264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60013084</td>\n",
       "      <td>29121</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Vary_doublet</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID  Genome      Pipeline version  \\\n",
       "0_0   20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_10  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_20  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "0_30  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "10_0  20220928-IGVF-D0  GRCh38  cellranger-arc-2.0.1   \n",
       "\n",
       "      Estimated number of cells  Feature linkages detected  Linked genes  \\\n",
       "0_0                        8631                      15570          5270   \n",
       "0_10                       8584                      13139          4687   \n",
       "0_20                       8571                      11718          4469   \n",
       "0_30                       8572                      10058          3964   \n",
       "10_0                       8594                      11004          4241   \n",
       "\n",
       "      Linked peaks  ATAC Confidently mapped read pairs  \\\n",
       "0_0          11303                              0.8748   \n",
       "0_10         10368                              0.8744   \n",
       "0_20          9926                              0.8749   \n",
       "0_30          9226                              0.8748   \n",
       "10_0          8788                              0.8710   \n",
       "\n",
       "      ATAC Fraction of genome in peaks  \\\n",
       "0_0                             0.0424   \n",
       "0_10                            0.0424   \n",
       "0_20                            0.0425   \n",
       "0_30                            0.0427   \n",
       "10_0                            0.0401   \n",
       "\n",
       "      ATAC Fraction of high-quality fragments in cells  ...  \\\n",
       "0_0                                             0.9335  ...   \n",
       "0_10                                            0.9264  ...   \n",
       "0_20                                            0.9246  ...   \n",
       "0_30                                            0.9224  ...   \n",
       "10_0                                            0.9264  ...   \n",
       "\n",
       "      GEX Reads mapped to genome  GEX Reads with TSO  \\\n",
       "0_0                       0.9981                 0.0   \n",
       "0_10                      0.9980                 0.0   \n",
       "0_20                      0.9981                 0.0   \n",
       "0_30                      0.9981                 0.0   \n",
       "10_0                      0.9981                 0.0   \n",
       "\n",
       "      GEX Sequenced read pairs  GEX Total genes detected  GEX Valid UMIs  \\\n",
       "0_0                   60314841                     29202             1.0   \n",
       "0_10                  60131721                     29242             1.0   \n",
       "0_20                  59914589                     29164             1.0   \n",
       "0_30                  60091431                     29190             1.0   \n",
       "10_0                  60013084                     29121             1.0   \n",
       "\n",
       "      GEX Valid barcodes          type  amb_rate  doublet_rate  num_donors  \n",
       "0_0                  1.0  Vary_doublet         0             0           4  \n",
       "0_10                 1.0  Vary_doublet         0            10           4  \n",
       "0_20                 1.0  Vary_doublet         0            20           4  \n",
       "0_30                 1.0  Vary_doublet         0            30           4  \n",
       "10_0                 1.0  Vary_doublet        10             0           4  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "dd34a3c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_summary.to_csv(f'{projdir}/csv/final_figures/tables/table_s4_sims_crarc_qc_lowcov.csv',\n",
    "                   sep=',', header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6b039d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scFates",
   "language": "python",
   "name": "scfates"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
